File: hooks/safety/git-safety-check.py
#!/usr/bin/env python3
"""
Git Safety Check Hook for Claude Code
Prevents dangerous Git operations and sensitive file commits

Protects UTXOracle from:
- Force push to main/master
- Deletion of protected branches
- Committing sensitive files (.env, bitcoin.conf, etc.)
- Large file commits (>50MB)
- Sensitive keywords in commit messages
"""

import json
import re
import subprocess
import sys
from pathlib import Path

# Configuration
PROTECTED_BRANCHES = ["main", "master", "production"]
LARGE_FILE_THRESHOLD = 50_000_000  # 50MB

# Sensitive file patterns (UTXOracle specific)
SENSITIVE_PATTERNS = [
    r"\.env$",  # Environment variables
    r"\.env\..*",  # .env.local, .env.production
    r"bitcoin\.conf$",  # Bitcoin RPC credentials
    r"\.cookie$",  # Bitcoin auth cookie
    r".*\.pem$",  # SSH keys
    r".*\.key$",  # Private keys
    r"id_rsa",  # SSH private key
    r"\.aws/credentials",  # AWS credentials (future)
    r"\.pypirc$",  # PyPI credentials
]

# Sensitive keywords in content/messages
SENSITIVE_KEYWORDS = [
    "BROWSERBASE_API_KEY",
    "BROWSERBASE_PROJECT_ID",
    "rpcuser",
    "rpcpassword",
    "private_key",
    "secret_key",
    "api_key",
    "password",
    "token",
]

# Bad commit message patterns (block these!)
BAD_COMMIT_PATTERNS = [
    r"\[⚠️\s*Agent:\s*unknown\]",  # Unknown agent - must identify
    r"PARTIAL",  # Incomplete commit - finish the work first
    r"^WIP$",  # Work in progress only - add description
    r"^fixup!?\s*$",  # Fixup without context
    r"^squash!?\s*$",  # Squash without context
]
# NOTE: AUTO-CHECKPOINT is allowed - it's automatic and serves safety purposes


def get_current_branch():
    """Get current Git branch name"""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"], capture_output=True, text=True, check=True
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError:
        return None


def is_protected_branch(branch):
    """Check if branch is protected"""
    return branch in PROTECTED_BRANCHES


def check_force_push(command):
    """Detect force push attempts"""
    return bool(("--force" in command or "-f" in command) and "push" in command)


def check_branch_deletion(command):
    """Detect branch deletion attempts"""
    if "branch" in command and ("-D" in command or "-d" in command):
        # Extract branch name
        for protected in PROTECTED_BRANCHES:
            if protected in command:
                return protected
    return None


def check_sensitive_files(command):
    """Check if commit includes sensitive files"""
    sensitive_files = []

    # Get list of files to be committed
    if "commit" in command or "add" in command:
        try:
            # Check staged files
            result = subprocess.run(
                ["git", "diff", "--cached", "--name-only"],
                capture_output=True,
                text=True,
                check=True,
            )
            staged_files = result.stdout.strip().split("\n")

            for file_path in staged_files:
                if not file_path:
                    continue

                # Check against sensitive patterns
                for pattern in SENSITIVE_PATTERNS:
                    if re.search(pattern, file_path, re.IGNORECASE):
                        sensitive_files.append(file_path)
                        break

        except subprocess.CalledProcessError:
            pass

    return sensitive_files


def check_large_files():
    """Check for large files in staging area"""
    large_files = []

    try:
        # Get staged files with sizes
        result = subprocess.run(["git", "diff", "--cached", "--name-only"], capture_output=True, text=True, check=True)
        staged_files = result.stdout.strip().split("\n")

        for file_path in staged_files:
            if not file_path or not Path(file_path).exists():
                continue

            file_size = Path(file_path).stat().st_size
            if file_size > LARGE_FILE_THRESHOLD:
                large_files.append({"file": file_path, "size": file_size, "size_mb": file_size / 1_000_000})

    except (subprocess.CalledProcessError, FileNotFoundError):
        pass

    return large_files


def check_commit_message(command):
    """Check commit message for sensitive keywords"""
    found_keywords = []

    if "commit" in command and "-m" in command:
        # Extract message from command
        message_match = re.search(r'-m\s+["\']([^"\']+)["\']', command)
        if message_match:
            message = message_match.group(1).lower()

            for keyword in SENSITIVE_KEYWORDS:
                if keyword.lower() in message:
                    found_keywords.append(keyword)

    return found_keywords


def check_bad_commit_patterns(command):
    """Check commit message for bad patterns (PARTIAL, unknown agent, etc.)"""
    if "commit" not in command:
        return None

    # Try to extract message from -m flag
    message_match = re.search(r'-m\s+["\']([^"\']+)["\']', command)
    if not message_match:
        # Try HEREDOC pattern
        message_match = re.search(r'-m\s+"\$\(cat <<[\'"]?EOF[\'"]?\n(.*?)\nEOF', command, re.DOTALL)

    if message_match:
        message = message_match.group(1)
        for pattern in BAD_COMMIT_PATTERNS:
            if re.search(pattern, message, re.IGNORECASE):
                return pattern, message[:100]

    return None


def main():
    """Main hook logic"""
    try:
        # Read input from Claude Code
        input_data = json.loads(sys.stdin.read())

        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        if not command or "git" not in command:
            # Not a git command, allow
            sys.exit(0)

        current_branch = get_current_branch()
        warnings = []
        errors = []

        # Check 1: Force push to protected branch
        if check_force_push(command):
            if current_branch and is_protected_branch(current_branch):
                errors.append(
                    f"❌ BLOCKED: Force push to protected branch '{current_branch}' is not allowed.\n"
                    f"   Use 'git push' without --force, or create a feature branch first."
                )
            else:
                warnings.append(
                    f"⚠️  WARNING: Force push detected on branch '{current_branch}'.\n"
                    f"   This may overwrite remote history. Proceed with caution."
                )

        # Check 2: Branch deletion
        deleted_branch = check_branch_deletion(command)
        if deleted_branch:
            errors.append(
                f"❌ BLOCKED: Cannot delete protected branch '{deleted_branch}'.\n"
                f"   Protected branches: {', '.join(PROTECTED_BRANCHES)}"
            )

        # Check 3: Sensitive files
        sensitive_files = check_sensitive_files(command)
        if sensitive_files:
            warnings.append(
                "⚠️  WARNING: Attempting to commit sensitive files:\n"
                + "\n".join(f"   - {f}" for f in sensitive_files)
                + "\n   Review carefully before proceeding. Consider adding to .gitignore."
            )

        # Check 4: Large files
        large_files = check_large_files()
        if large_files:
            warnings.append(
                "⚠️  WARNING: Large files detected in commit:\n"
                + "\n".join(f"   - {f['file']} ({f['size_mb']:.1f} MB)" for f in large_files)
                + f"\n   Consider using Git LFS for files >{LARGE_FILE_THRESHOLD / 1_000_000:.0f}MB."
            )

        # Check 5: Sensitive keywords in commit message
        sensitive_keywords = check_commit_message(command)
        if sensitive_keywords:
            warnings.append(
                "⚠️  WARNING: Sensitive keywords in commit message:\n"
                + "\n".join(f"   - {k}" for k in sensitive_keywords)
                + "\n   Avoid committing secrets in messages."
            )

        # Check 6: Bad commit message patterns (PARTIAL, unknown agent, etc.)
        bad_pattern = check_bad_commit_patterns(command)
        if bad_pattern:
            pattern, message_preview = bad_pattern
            errors.append(
                f"❌ BLOCKED: Bad commit message pattern detected.\n"
                f"   Pattern: {pattern}\n"
                f"   Message: {message_preview}...\n"
                f"   Use a descriptive commit message following conventional commits."
            )

        # If errors, block the operation
        if errors:
            output = {
                "hookSpecificOutput": {
                    "hookEventName": "PreToolUse",
                    "shouldBlock": True,
                    "blockMessage": "\n\n".join(errors),
                }
            }
            print(json.dumps(output))
            sys.exit(1)

        # If warnings, show but allow
        if warnings:
            output = {
                "hookSpecificOutput": {
                    "hookEventName": "PreToolUse",
                    "shouldBlock": False,
                    "message": "\n\n".join(warnings),
                }
            }
            print(json.dumps(output))
            sys.exit(0)

        # All clear
        sys.exit(0)

    except json.JSONDecodeError:
        # Not valid JSON input, allow operation
        sys.exit(0)
    except Exception as e:
        # On error, fail open (allow operation but log)
        print(
            json.dumps(
                {
                    "hookSpecificOutput": {
                        "hookEventName": "PreToolUse",
                        "message": f"⚠️  Git safety check error: {str(e)}",
                    }
                }
            )
        )
        sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/safety/git-safety-check.js
#!/usr/bin/env node
/**
 * Git Safety Check Hook for Claude Code
 * Prevents dangerous Git operations
 *
 * Protects from:
 * - Force push to main/master
 * - git reset --hard
 * - git clean -f
 * - git checkout . (discards all changes)
 * - git branch -D (force delete)
 *
 * Returns: { decision: "block", reason: "..." } or {} to allow
 */

const path = require('path');
const { readStdinJson, output, runCommand } = require(path.join(__dirname, '..', '..', 'lib', 'utils.js'));

// Configuration
const PROTECTED_BRANCHES = ['main', 'master', 'production'];

/**
 * Get current Git branch name
 */
function getCurrentBranch() {
  const result = runCommand('git rev-parse --abbrev-ref HEAD');
  return result.success ? result.output : null;
}

/**
 * Check if branch is protected
 */
function isProtectedBranch(branch) {
  return PROTECTED_BRANCHES.includes(branch);
}

/**
 * Check for force push attempts
 */
function checkForcePush(command) {
  // Patterns for force push
  const forcePushPatterns = [
    /git\s+push\s+.*--force/,
    /git\s+push\s+.*\s-f(?:\s|$)/,
    /git\s+push\s+--force-with-lease/  // Safer but still risky
  ];

  for (const pattern of forcePushPatterns) {
    if (pattern.test(command)) {
      // Check if pushing to main/master specifically
      if (/\s+(origin\s+)?(main|master)(\s|$)/.test(command)) {
        return { isForce: true, toProtected: true };
      }
      return { isForce: true, toProtected: false };
    }
  }
  return { isForce: false, toProtected: false };
}

/**
 * Check for hard reset
 */
function checkHardReset(command) {
  return /git\s+reset\s+--hard/.test(command);
}

/**
 * Check for git clean with force
 */
function checkGitClean(command) {
  return /git\s+clean\s+.*(--force|-[a-z]*f)/.test(command);
}

/**
 * Check for checkout . (discard all changes)
 */
function checkCheckoutDiscard(command) {
  return /git\s+checkout\s+\./.test(command) || /git\s+restore\s+\./.test(command);
}

/**
 * Check for force branch deletion
 */
function checkBranchForceDelete(command) {
  // Check ALL segments of chained commands
  const segments = command.split(/\s*(?:&&|\|\||;)\s*/);
  for (const segment of segments) {
    const match = segment.trim().match(/git\s+branch\s+.*-D\s+(.+)/);
    if (match) {
      const branches = match[1].trim().split(/\s+/)
        .map(b => b.replace(/["']/g, ''));
      const protectedFound = branches.filter(b => PROTECTED_BRANCHES.includes(b));
      return {
        isForceDelete: true,
        isProtected: protectedFound.length > 0,
        branchName: branches.join(', '),
        protectedBranches: protectedFound
      };
    }
  }
  return { isForceDelete: false };
}

async function main() {
  try {
    const input = await readStdinJson();

    const toolName = input.tool_name || '';
    const toolInput = input.tool_input || {};
    const command = toolInput.command || '';

    // Only check Bash commands with git
    if (toolName !== 'Bash' || !command.includes('git')) {
      output({});
      process.exit(0);
    }

    const currentBranch = getCurrentBranch();
    const errors = [];
    const warnings = [];

    // Check 1: Force push
    const forcePushResult = checkForcePush(command);
    if (forcePushResult.isForce) {
      if (forcePushResult.toProtected) {
        errors.push(
          `BLOCKED: Force push to protected branch (main/master) is not allowed.\n` +
          `Use 'git push' without --force, or create a feature branch first.`
        );
      } else if (currentBranch && isProtectedBranch(currentBranch)) {
        errors.push(
          `BLOCKED: Force push to protected branch '${currentBranch}' is not allowed.\n` +
          `Use 'git push' without --force, or switch to a feature branch.`
        );
      } else {
        warnings.push(
          `WARNING: Force push detected on branch '${currentBranch || 'unknown'}'.\n` +
          `This may overwrite remote history. Proceed with caution.`
        );
      }
    }

    // Check 2: Hard reset
    if (checkHardReset(command)) {
      errors.push(
        `BLOCKED: 'git reset --hard' will discard all uncommitted changes.\n` +
        `Use 'git stash' first to save changes, or 'git reset --soft' to keep them.`
      );
    }

    // Check 3: Git clean with force
    if (checkGitClean(command)) {
      errors.push(
        `BLOCKED: 'git clean -f' will permanently delete untracked files.\n` +
        `Use 'git clean -n' first to see what would be deleted.`
      );
    }

    // Check 4: Checkout discard
    if (checkCheckoutDiscard(command)) {
      errors.push(
        `BLOCKED: 'git checkout .' or 'git restore .' will discard all uncommitted changes.\n` +
        `Use 'git stash' first to save changes.`
      );
    }

    // Check 5: Force branch deletion
    const branchDeleteResult = checkBranchForceDelete(command);
    if (branchDeleteResult.isForceDelete) {
      if (branchDeleteResult.isProtected) {
        errors.push(
          `BLOCKED: Cannot force-delete protected branch '${branchDeleteResult.branchName}'.\n` +
          `Protected branches: ${PROTECTED_BRANCHES.join(', ')}`
        );
      } else {
        warnings.push(
          `WARNING: Force deleting branch '${branchDeleteResult.branchName}'.\n` +
          `This cannot be undone if the branch is not merged. Consider using -d instead.`
        );
      }
    }

    // If errors, block the operation
    if (errors.length > 0) {
      output({
        decision: 'block',
        reason: errors.join('\n\n')
      });
      process.exit(0);
    }

    // If warnings, show but allow (via empty object - Claude Code handles warnings differently)
    if (warnings.length > 0) {
      // For now, just allow with warnings logged to stderr
      console.error(warnings.join('\n\n'));
    }

    // All clear
    output({});
    process.exit(0);

  } catch (err) {
    // On error, fail open (allow operation)
    console.error(`Git safety check error: ${err.message}`);
    output({});
    process.exit(0);
  }
}

main();
